### PR TITLE
Signals

### PR DESCRIPTION
**Pull request by EPashkin**
Part of #199
WIP

I don't planing read these signal attributes as don't see its use now: when, detailed, action, no-recurse, introspectable, no-hooks

### CODE REVIEW

**Comment by gkoz** [comment]
So there is `no-recurse`! This is nice. Such signals could take `FnMut` instead of `Fn`.

File: src/parser.rs [hunk 1]
```diff
@@ -13,6 +13,8 @@ use library::*;
 type Reader = EventReader<BufReader<File>>;
 type Attributes = Vec<OwnedAttribute>;
 
+const EMPTY_CTYPE: &'static str = "/*EMPTY*/";
```
Comment (gkoz) [inline src/parser.rs]: This is obviously suboptimal but I'm not sure if it's better to have two `Parameter` types or put up with optional `c_type`. Perhaps the composition&`Deref` trick could help here too.

Comment (EPashkin) [inline src/parser.rs]: Seems for signals separate classes not needed.
Used 111 types can has "object" c_type, and can filled with it (like `GdkWindow*`) , others 5 is array types and can be processed later.
Full list https://gist.github.com/EPashkin/b13bf3e8e4a6087a8078.
Seems no primitive types without c:type in signals.

**Comment by EPashkin** [comment]
Add signals empty ctypes post detecting

**Comment by EPashkin** [comment]
Made `enumerate` but can't do single pass update due needed `Library::type_` in `c_type_by_type_id`

**Comment by EPashkin** [comment]
Rebased

**Comment by EPashkin** [comment]
@gkoz There really needed type bounds in trampolines?
Its seems unused, and if I remove it in https://github.com/gtk-rs/gtk/blob/master/src/signal.rs#L1041-L1060 `gtk` still builds fine.

**Comment by gkoz** [comment]
Uh-oh. This looks like a giant bug: `T` is not used at all! I think it was supposed to be more like this
```rust
    unsafe extern "C" fn void_trampoline<T>(this: *mut GtkComboBox, f: &Box<Fn(&T) + 'static>)
    where T: IsA<ComboBox> {
        callback_guard!();
        f(&ComboBox::from_glib_none(this).downcast_unchecked());
    }
```

**Comment by gkoz** [comment]
We could make signals first-class types kinda like they used to be in rgtk.

```
unsafe trait Signal<Receiver> {
    type Closure;
    /// ...
}

enum ButtonClicked { }

unsafe impl<O: IsA<Button>> Signal<O> for ButtonClicked {
    type Closure = Fn(&O);
    // ...
}

trait ObjectExt {
    fn connect<S: Signal<Self>>(&self, f: S::Closure) -> u64;
    fn connect_after<S: Signal<Self>>(&self, f: S::Closure) -> u64;
}

button.connect::<ButtonClicked>(|btn| do_something());
```

**Comment by EPashkin** [comment]
IMHO worth to try.

**Comment by gkoz** [comment]
My example wouldn't compile though because `S::Closure` is unsized :( It has to be more complex.

**Comment by GuillaumeGomez** [comment]
Any news on this @EPashkin?

**Comment by EPashkin** [comment]
Sorry, no free time.
Anyway IMHO it don't need in next crate release as will subject of changes.

**Comment by EPashkin** [comment]
Ready for merging.
Some part is not done, but I prefer do it in separate PRs:
1) Versions
2) Configuration (ignore)
3) Remove doubleboxing

**Comment by GuillaumeGomez** [comment]
I think it's better to stop here, it's already a very big PR! Thanks a lot for your work @EPashkin!

**Comment by EPashkin** [comment]
Strange, my generator don't want generate `MenuItem::connect_toggle_size_request`, I don't understand why.

**Comment by EPashkin** [comment]
Now understand, it accept `gpointer` in Gtk-3.0.gir, so we don't know real type.
In https://github.com/gtk-rs/gtk/pull/300 states that it also `int`, not sure that it right.

**Comment by EPashkin** [comment]
https://people.gnome.org/~shaunm/girdoc/C/Gtk.MenuItem.toggle_size_request.html
states that it `int*` not `int`

**Comment by EPashkin** [comment]
@GuillaumeGomez  C-source states also, so it need be fixed (or disabled) in your PR.

**Comment by EPashkin** [comment]
Minor changes in `codegen::signal::bounds` (removed if in match)

**Comment by GuillaumeGomez** [comment]
I'll rmeove them from my PR.

**Comment by gkoz** [comment]
When trying this on `gio` looks like some imports are missing. Funny thing about `gio`: I've only just noticed that we'll need an override to force generation of traits: `Gio-2.0.gir` can't know that `Gtk-3.0.gir` subclasses `Application`.

Do you think we should keep adding the `connect_` methods for now rather than design a generic API?

**Comment by EPashkin** [comment]
Oh, I fully forgot about imports, so it may has problem even in gtk :(
Planned fixing imports after add version and ignore, when updating gtk.

`connect_` has good use with closures. Do not see how it can be replaced.

**Comment by gkoz** [comment]
I'm gravitating toward something like
```rust
widget.connect::<WidgetDeleteEvent>(|this, event| { ... });
```
(eventually `widget.connect::<Widget::DELETE_EVENT>(...);`)

which could take several closure kinds:
```rust
Fn(&T, &Event) -> Inhibit + 'static
Fn(&T, &Event) + 'static
// possibly
Fn(&Event) -> Inhibit + 'static
Fn(&Event) + 'static
// maybe even?
Fn() -> Inhibit + 'static
Fn() + 'static // this one would allow `window.connect::<WidgetDeleteEvent>(gtk::main_quit)`
```

But this might not get fleshed out in a while.

**Comment by EPashkin** [comment]
IMHO approach with `::<` will bad for IDE's (racer) autocomplete function.
Secondly we get some "God" function that also not good.

**Comment by EPashkin** [comment]
Imports is sure problem so this PR still WIP

**Comment by EPashkin** [comment]
Type `GtkBox` conflict with signals in almost all files, can I rename its generated type to 'Box_'?
Or better import `std::boxed::Box` as `Box_` and generate signal with it?

**Comment by gkoz** [comment]
> Type GtkBox conflict with signals in almost all files, can I rename its generated type to Box_?
Or better import std::boxed::Box as Box_ and generate signal with it?

Tough one :( Is it feasible to avoid importing just this one type and always spell it as `::Box`?

**Comment by EPashkin** [comment]
`::Box` work same as just `Box` from gtk. `std::boxed::Box` IMHO too long and working only after `use std;`

**Comment by gkoz** [comment]
If there's no `use Box;`, I believe `Box` should resolve to `::std::boxed::Box` and `::Box` to `::auto::box_::Box`. But this special case it could be too complex to implement.

**Comment by EPashkin** [comment]
You right about both parts: it work but unnecessary complex

**Comment by EPashkin** [comment]
Currently I stay with `use std::boxed::Box as Box_`

**Comment by gkoz** [comment]
Alright, let's look into renaming our `Box` to `Box_`. Reexporting it as `Box` manually should avoid breaking any code.

**Comment by gkoz** [comment]
For reference, I've outlined the way to avoid double-boxing [here](https://github.com/gkoz/gtk-rs-glib/commit/628c541d02e09a37a3282998e3a05319c6060b19) but the significance of the monomorphization cost (how bad it is to generate a copy of each trampoline per closure) is not clear to me and in the best case this feels like premature optimization.

**Comment by EPashkin** [comment]
If we reexporing gtk's `Box` as `Box` to users, then I don't see any reason to rename it internally.

**Comment by EPashkin** [comment]
I currently don't see any problem with many copies for trampolines, this too small anyway.
Also I decide not combine trampolines and generate one trampoline for each signal.

We can do `double-boxing` fix after finishing applying signal generation to Gtk.

**Comment by gkoz** [comment]
> If we reexporing gtk's Box as Box to users, then I don't see any reason to rename it internally.

I was thinking of trying to keep the code idiomatic. Signal bindings are the most fragile it seems and putting speed bumps like `Box_` or `::std::boxed::Box` into that code could make reviewing harder. There's value in everyone knowing that `Box` is what they expect.
At the same time users would have a choice of using `Box_` or `Box` whichever suits them best. But having two names for the struct could be confusing in its own way I suppose.

**Comment by EPashkin** [comment]
Two Boxes mixed almost only in generated code, so IMHO review or Box_\Box dilemma its not problem.

**Comment by gkoz** [comment]
I don't feel strongly about this so okay, let's go with your solution.

File: src/analysis/signals.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    let trampoline_name = trampolines::analyze(env, signal, type_tid, in_trait, trampolines,
+                                               &mut used_types, version);
+
+    if trampoline_name.is_some() {
+        for s in used_types {
+            if let Some(i) = s.find("::") {
+                imports.add(&s[..i], version);
+            } else {
+                imports.add(&s, version);
+            }
```
Comment (gkoz) [inline src/analysis/signals.rs]: There should probably be an `Imports` method, that does this.

Comment (EPashkin) [inline src/analysis/signals.rs]: You right, this logic better be in one place, even if only 2 uses.

File: src/codegen/trampoline.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    } else {
+        ("", " {")
+    };
+
+    let params_str = trampoline_parameters(env, analysis);
+    let func_str = func_string(env, analysis, None::<(&str, &str)>);
+    let ret_str = trampoline_returns(env, analysis);
+
+    try!(version_condition(w, env, analysis.version, false, 0));
+    try!(writeln!(w, "unsafe extern \"C\" fn {}{}({}, f: &Box_<{}>){}{}",
```
Comment (gkoz) [inline src/codegen/trampoline.rs]: I'd rather we continued to clean up the extern fns' signatures, put a `gpointer` here and cast it in the body.

Comment (EPashkin) [inline src/codegen/trampoline.rs]: Safer do it after gtk has current variant.
Also I plan use `Chunks` here too, so this will be reworked anyway.

**Comment by EPashkin** [comment]
Now we got working version.
Gtk update coming soon.
There still some errors in gtk, I solve it ignoring some signals,
also add issues latter (or update #199)

**Comment by EPashkin** [comment]
Rebased on skip_assert fix

**Comment by gkoz** [comment]
I've only realized this now, almost too late: the generator doesn't know when to translate `bool` to `Inhibit` does it?

**Comment by EPashkin** [comment]
Yes, I don't think about it.

**Comment by EPashkin** [comment]
You mean `Inhibit` in closure return to `gboolean` in trampoline return?

**Comment by gkoz** [comment]
Yes, e.g. generating `RangeExt::connect_change_value` would be a breaking change.

**Comment by EPashkin** [comment]
Added issue, remove generation for Range

**Comment by EPashkin** [comment]
AppVeyor build succeeded

**Comment by EPashkin** [comment]
By way anyone know how do "Hooray" emoji?

**Comment by mitaa** [comment]
http://www.emoji-cheat-sheet.com/

`:tada:`

:tada:

**Comment by EPashkin** [comment]
Thanks. Just don't see it in github help.

**Pull request merged**